extra_data/reader.py

```diff
@@ -708,7 +712,50 @@ def select(self, seln_or_source_glob, key_glob='*'):
         files = [f for f in self.files
                  if f.all_sources.intersection(selection.keys())]
 
-        return DataCollection(files, selection=selection, train_ids=self.train_ids)
+        if require_all:
+            # Select only those trains for which all selected sources
+            # and keys have data, i.e. have a count > 0 in their
+            # respective INDEX section.
+
+            train_ids = self.train_ids
+
+            for source, keys in selection.items():
+                if source in self.instrument_sources:
+                    # For INSTRUMENT sources, the INDEX is saved by
+                    # key group, which is the first hash component. In
+                    # practice, this is always 'data', but this is not
+                    # guaranteed and there may be more than one.
+                    if keys is None:
+                        # All keys are selected.
+                        keys = self.keys_for_source(source)
+
+                    groups = {key.partition('.')[0] for key in keys}
+                else:
+                    # The index for CONTROL data may be assumed to not
+                    # depend on the key, but only the source.
+                    groups = ['']
+
+                source_tids = []
+
+                for f, group in product(self._source_index[source], groups):
+                    # Add the trains with data found for each key group
+                    # in each file.
+                    source_tids = np.union1d(
+                        f.train_ids[f.get_index(source, group)[1] > 0],
+                        source_tids)
+
+                # Remove any trains previously selected, for which this
+                # selected source had no data.
+                train_ids = np.intersect1d(train_ids, source_tids)
```

Comment:
I think we should intersect the different key groups within a source, rather than unioning them. If you've selected keys in two groups from the same source, you want `require_all=True` to select only trains where all of those keys have data.

With the DAQ writing the data, I think all key groups in a source will have data for the same trains, so this doesn't matter. But it's not a property of the file structure, and data reduction might not preserve that property, so we may as well anticipate the possibility.

Reply:
Hmmm, I fully agree, not sure what I was thinking here. A union makes no sense, as it may include trains where one or more selected key group are missing data. It may be a remnant of a previous attempt using sets and I misinterpreted the binary operator.